sdk/servicebus/service-bus/src/core/messageReceiver.ts

```diff
@@ -1025,29 +1021,70 @@ export class MessageReceiver extends LinkEntity {
         };
         return reject(translate(e));
       }, Constants.defaultOperationTimeoutInSeconds * 1000);
+
+      // This takes care of resolving the promise by waiting on the onSettled event
       this._deliveryDispositionMap.set(delivery.id, {
         resolve: resolve,
         reject: reject,
         timer: timer
       });
-      if (operation === DispositionType.complete) {
-        delivery.accept();
-      } else if (operation === DispositionType.abandon) {
-        const params: any = {
-          undeliverable_here: false
-        };
-        if (options.propertiesToModify) params.message_annotations = options.propertiesToModify;
-        delivery.modified(params);
-      } else if (operation === DispositionType.defer) {
-        const params: any = {
-          undeliverable_here: true
-        };
-        if (options.propertiesToModify) params.message_annotations = options.propertiesToModify;
-        delivery.modified(params);
-      } else if (operation === DispositionType.deadletter) {
-        delivery.reject(options.error || {});
+
+      try {
+        if (operation === DispositionType.complete) {
+          delivery.accept();
+        } else if (operation === DispositionType.abandon) {
+          const params: any = {
+            undeliverable_here: false
+          };
+          if (options.propertiesToModify) params.message_annotations = options.propertiesToModify;
+          delivery.modified(params);
+        } else if (operation === DispositionType.defer) {
+          const params: any = {
+            undeliverable_here: true
+          };
+          if (options.propertiesToModify) params.message_annotations = options.propertiesToModify;
+          delivery.modified(params);
+        } else if (operation === DispositionType.deadletter) {
+          delivery.reject(options.error || {});
+        }
+      } catch (err) {
+        log.error(
+          `[%s] Operation "%s" failed on message with delivery id: %d, with error: %0`,
+          this._context.namespace.connectionId,
+          operation,
+          message.delivery.id,
+          err
+        );
+        throw err;
+      } finally {
+        this._clearMessageLockRenewTimer(message.messageId as string);
```

Comment:
Moving the `_clearMessageLockRenewTimer()` here results in clearing the lock renew timer as soon as the request to settle the message is fired instead of waiting for the message settlement to succeed or fail

Reply:
Yes, that was the intent. But teh original implementation doesn't waitto clear timer on the `onSettled` event, so I thought the assumption was that start of a settle operation would mark end of lock renewal.
But with retries, it makes sense to do it at the end

Reply:
Fair point, I misunderstood the intent of the timer.

With this move, we might end up in two unexpected cases of lock renewal and message settlement processes conflicting.

Case 1:
- user calls complete() on a message
- service settles the message and the onSettled event gets fired
- the timer on lock renewal completes, the event loop is updated with instructions to renew the lock
- `_clearMessageLockRenewTimer()` is run, but the request to renew lock is already queued up.
- Request to renew lock fails as the message is already settled

Case 2:
- user calls complete() on a message
- the timer on lock renewal completes, the event loop is updated with instructions to renew the lock
- service settles the message and the _onSettled event gets fired
- Request to renew lock fails as the message is already settled
- `_clearMessageLockRenewTimer()` is run

Reply:
The onSettled handler clears the timer on success, so these cases shouldn't be a problem.

In case 1, step 3 will not happen and for case 2, step  wouldn't happen as the timer would have been cleared. 
About the general possibility of renew lock failing due to a race condition, I don't think this is avoidable? Since we already do our best to clear the lock renewal timer when settling concludes.

Reply:
Been looking into it some more and agree that this race condition is a rare but possible scenario. The work around updated in PR is to not perform autolockrenewal when a message is already settled.

Discussed with @AlexGhiondea as well and we found a different bug/scenario where if a lock renewal fails silently in background for any reason in case of sessions, the error is currently being swallowed.
And the user may subsequently see a "SessionLockLostError" while performing message/client operations. Which would come as a surprise if auto lock renewal is expected to be in progress. Just like for non-session receiver/entities, updated session receivers to correctly handle and surface the error.